# 1st Delievery - EDA

# Table of Contents

<a class="anchor" id="top"></a>


1. [Context & Importing Libraries & Data](#1.Context-Importing-Libraries-&-Data)    

   1.1 [Context](#1.1-Context)
   
   1.2 [Loading and Reading the Data](#1.2Loading-and-Reading-the-Data)
   
   

    
    
3. [Exploratory Data Analysis](#2.-Exploratory-Data-Analysis)

    2.1 [Incoherencies](#2.1-Incoherencies)

    2.2 [Unique Feature-Pair Analysis ](#2.2-Unique-Feature-Pair-Analysis)

    2.3 [Duplicates](#2.3-Duplicates)
    
    2.4 [Missing Values](#2.4-Missing-Values)
    
    2.5 [Outliers](#2.5-Outliers)
    
    2.6 [Feature Engineering](#2.6-Feature-Engineering)
        
4. [Explorating Data Vizualizations](#3.-Explorating-Data-Vizualizations)

Dont follow this for now 

# 1. Context & Importing Libraries & Data

## 1.1. Context
We are consultants for ABCDEats Inc (ABCDE), a food delivery service with a vast range of restaurants with diferent cuisines. We are going to analyse tree months of data from tree cities to provide  a data-driven strategy than englobes tree different segmentation perpectives. 

## 1.2 Loading and Reading the Data

### Importing Libraries

In [1]:
# All our libraries
import sqlite3
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from math import ceil

# for better resolution plots
%config InlineBackend.figure_format = 'retina' # optionally, you can change 'svg' to 'retina'. 
#retina is for better resolutions in images 

# Seeting seaborn style
sns.set()

## Metadata
`customer_id` - Unique identifier for each customer <br>
`customer_region` - Geographic region where the customer is located <br>
`customer_age` - Age of the customer <br>
`vendor_count` - Number of unique vendors the customer has ordered from <br>
`product_count` - Total number of products the customer has ordered <br>
`is_chain` - Indicates whether the customer’s order was from a chain restaurant <br>
`first_order` - Number of days from the start of the dataset when the customer first placed an order. <br>
`last_order` - Number of days from the start of the dataset when the customer most recently placed an order. <br>
`last_promo` - The category of the promotion or discount most recently used by the customer. <br>
`payment_method` - Method most recently used by the customer to pay for their orders. <br>
`CUI_(cuisine)`- The amount in monetary units spent by the customer from the indicated type of cuisine. <br>
- CUI_Beverages, CUI_Cafe', 'CUI_Chicken Dishes', 'CUI_Chinese', 'CUI_Desserts','CUI_Healthy', 'CUI_Indian', 'CUI_Italian', 'CUI_Japanese', 'CUI_Noodle Dishes', 'CUI_OTHER', 'CUI_Street Food / Snacks', CUI_Thai <br>

`DOW_0 to DOW_6` - Number of orders placed on each day of the week (0 = Sunday, 6 = Saturday). <br>
`HR_0 to HR_23` - Number of orders placed during each hour of the day (0 = midnight, 23 = 11 PM). <br>
ders.



### Reading the Data

In [2]:
#understanding our directory
print(os.listdir())

['.ipynb_checkpoints', 'Data_explo_structered.ipynb', 'DM2425_ABCDEats_DATASET.csv', 'First Part.ipynb', 'Other projects', 'Words_Pfds Documents']


In [3]:
ABCDE = pd.read_csv('DM2425_ABCDEats_DATASET.csv', sep = ',')   #sep is good to seperate data, we are specifying the delimeter
pd.set_option('display.max_columns', None) #to be able too see all columns
ABCDE.head(5) #checking if it is working correctly

customer_id customer_region  customer_age  vendor_count  product_count  \
0  1b8f824d5e            2360          18.0             2              5   
1  5d272b9dcb            8670          17.0             2              2   
2  f6d1b2ba63            4660          38.0             1              2   
3  180c632ed8            4660           NaN             2              3   
4  4eb37a6705            4660          20.0             2              5   

   is_chain  first_order  last_order last_promo payment_method  CUI_American  \
0         1          0.0           1   DELIVERY           DIGI          0.00   
1         2          0.0           1   DISCOUNT           DIGI         12.82   
2         2          0.0           1   DISCOUNT           CASH          9.20   
3         1          0.0           2   DELIVERY           DIGI          0.00   
4         0          0.0           2          -           DIGI         14.57   

   CUI_Asian  CUI_Beverages  CUI_Cafe  CUI_Chicken Dishes  CUI_Chinese  \
0       0.00            0.0       0.0                 0.0          0.0   
1       6.39            0.0       0.0                 0.0          0.0   
2       0.00            0.0       0.0                 0.0          0.0   
3      13.70            0.0       0.0                 0.0          0.0   
4      40.87            0.0       0.0                 0.0          0.0   

   CUI_Desserts  CUI_Healthy  CUI_Indian  CUI_Italian  CUI_Japanese  \
0           0.0          0.0       28.88          0.0           0.0   
1           0.0          0.0        0.00          0.0           0.0   
2           0.0          0.0        0.00          0.0           0.0   
3           0.0          0.0       17.86          0.0           0.0   
4           0.0          0.0        0.00          0.0           0.0   

   CUI_Noodle Dishes  CUI_OTHER  CUI_Street Food / Snacks  CUI_Thai  DOW_0  \
0                0.0        0.0                       0.0       0.0      1   
1                0.0        0.0                       0.0       0.0      1   
2                0.0        0.0                       0.0       0.0      1   
3                0.0        0.0                       0.0       0.0      0   
4                0.0        0.0                       0.0       0.0      0   

   DOW_1  DOW_2  DOW_3  DOW_4  DOW_5  DOW_6  HR_0  HR_1  HR_2  HR_3  HR_4  \
0      0      0      0      0      0      1   0.0     0     0     0     0   
1      0      0      0      0      0      1   0.0     0     0     0     0   
2      0      0      0      0      0      1   0.0     0     0     0     0   
3      1      0      0      0      0      1   0.0     0     0     0     0   
4      1      0      0      0      0      1   0.0     0     0     0     0   

   HR_5  HR_6  HR_7  HR_8  HR_9  HR_10  HR_11  HR_12  HR_13  HR_14  HR_15  \
0     0     0     0     0     0      0      0      0      0      0      0   
1     0     0     0     0     0      1      1      0      0      0      0   
2     0     0     0     0     1      0      1      0      0      0      0   
3     0     0     0     0     0      0      1      0      0      1      0   
4     0     0     0     1     1      0      0      0      0      0      0   

   HR_16  HR_17  HR_18  HR_19  HR_20  HR_21  HR_22  HR_23  
0      0      0      2      0      0      0      0      0  
1      0      0      0      0      0      0      0      0  
2      0      0      0      0      0      0      0      0  
3      0      0      0      0      0      0      0      0  
4      0      0      0      0      0      0      0      0

### Any footnotes?

In [4]:
ABCDE.tail(5)
# No, there isn't any

customer_id customer_region  customer_age  vendor_count  product_count  \
31883  f4e366c281            8670          30.0             1              1   
31884  f6b6709018            8670           NaN             1              1   
31885  f74ad8ce3f            8670          24.0             1              1   
31886  f7b19c0241            8670          34.0             1              1   
31887  fd40d3b0e0            4660          30.0             1              1   

       is_chain  first_order  last_order last_promo payment_method  \
31883         1          NaN           0    FREEBIE           DIGI   
31884         0          NaN           0   DELIVERY           DIGI   
31885         1          NaN           0          -           DIGI   
31886         0          NaN           0   DISCOUNT           DIGI   
31887         0          NaN           0          -           CASH   

       CUI_American  CUI_Asian  CUI_Beverages  CUI_Cafe  CUI_Chicken Dishes  \
31883          0.00       0.00          18.04       0.0                 0.0   
31884         18.04       0.00           0.00       0.0                 0.0   
31885          0.00       0.00           0.00       0.0                 0.0   
31886          0.00      12.03           0.00       0.0                 0.0   
31887          0.00       0.00           0.00       0.0                 0.0   

       CUI_Chinese  CUI_Desserts  CUI_Healthy  CUI_Indian  CUI_Italian  \
31883          0.0          0.00          0.0         0.0         0.00   
31884          0.0          0.00          0.0         0.0         0.00   
31885          0.0         17.79          0.0         0.0         0.00   
31886          0.0          0.00          0.0         0.0         0.00   
31887          0.0          0.00          0.0         0.0         7.91   

       CUI_Japanese  CUI_Noodle Dishes  CUI_OTHER  CUI_Street Food / Snacks  \
31883           0.0                0.0        0.0                       0.0   
31884           0.0                0.0        0.0                       0.0   
31885           0.0                0.0        0.0                       0.0   
31886           0.0                0.0        0.0                       0.0   
31887           0.0                0.0        0.0                       0.0   

       CUI_Thai  DOW_0  DOW_1  DOW_2  DOW_3  DOW_4  DOW_5  DOW_6  HR_0  HR_1  \
31883       0.0      0      0      0      0      0      0      1   0.0     0   
31884       0.0      0      0      0      0      0      0      1   0.0     0   
31885       0.0      0      0      0      0      0      0      1   0.0     0   
31886       0.0      0      0      0      0      0      0      1   0.0     0   
31887       0.0      0      0      0      0      0      0      1   0.0     0   

       HR_2  HR_3  HR_4  HR_5  HR_6  HR_7  HR_8  HR_9  HR_10  HR_11  HR_12  \
31883     0     0     0     0     0     0     0     0      0      0      0   
31884     0     0     0     0     0     0     0     0      0      0      0   
31885     0     0     0     1     0     0     0     0      0      0      0   
31886     0     0     0     0     0     0     0     0      0      1      0   
31887     0     0     0     0     0     0     0     0      0      0      0   

       HR_13  HR_14  HR_15  HR_16  HR_17  HR_18  HR_19  HR_20  HR_21  HR_22  \
31883      1      0      0      0      0      0      0      0      0      0   
31884      0      0      0      1      0      0      0      0      0      0   
31885      0      0      0      0      0      0      0      0      0      0   
31886      0      0      0      0      0      0      0      0      0      0   
31887      0      0      0      1      0      0      0      0      0      0   

       HR_23  
31883      0  
31884      0  
31885      0  
31886      0  
31887      0

### Preliminar notes
customer region -> there are 3 locations, for each city, can we trace a region code per city?

is_chain -> should return a binary result of 0 or 1. replace any value greater than 1 with 1, so that 1 returns "at least 1"

first order -> 

last_promo -> "None" = we can interpret non existent results as "the customer never had a promotion"

HR_0 to HR_23 -> set intervals to fit the data or only display values 



# 2. Exploratory Data Analysis

## 1.2 Data Description

In [5]:
# Check the shape
ABCDE.shape

(31888, 56)

In [6]:
ABCDE.columns

Index(['customer_id', 'customer_region', 'customer_age', 'vendor_count',
       'product_count', 'is_chain', 'first_order', 'last_order', 'last_promo',
       'payment_method', 'CUI_American', 'CUI_Asian', 'CUI_Beverages',
       'CUI_Cafe', 'CUI_Chicken Dishes', 'CUI_Chinese', 'CUI_Desserts',
       'CUI_Healthy', 'CUI_Indian', 'CUI_Italian', 'CUI_Japanese',
       'CUI_Noodle Dishes', 'CUI_OTHER', 'CUI_Street Food / Snacks',
       'CUI_Thai', 'DOW_0', 'DOW_1', 'DOW_2', 'DOW_3', 'DOW_4', 'DOW_5',
       'DOW_6', 'HR_0', 'HR_1', 'HR_2', 'HR_3', 'HR_4', 'HR_5', 'HR_6', 'HR_7',
       'HR_8', 'HR_9', 'HR_10', 'HR_11', 'HR_12', 'HR_13', 'HR_14', 'HR_15',
       'HR_16', 'HR_17', 'HR_18', 'HR_19', 'HR_20', 'HR_21', 'HR_22', 'HR_23'],
      dtype='object')

In [7]:
ABCDE.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31888 entries, 0 to 31887
Data columns (total 56 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customer_id               31888 non-null  object 
 1   customer_region           31888 non-null  object 
 2   customer_age              31161 non-null  float64
 3   vendor_count              31888 non-null  int64  
 4   product_count             31888 non-null  int64  
 5   is_chain                  31888 non-null  int64  
 6   first_order               31782 non-null  float64
 7   last_order                31888 non-null  int64  
 8   last_promo                31888 non-null  object 
 9   payment_method            31888 non-null  object 
 10  CUI_American              31888 non-null  float64
 11  CUI_Asian                 31888 non-null  float64
 12  CUI_Beverages             31888 non-null  float64
 13  CUI_Cafe                  31888 non-null  float64
 14  CUI_Ch

#### Observations: 
- maybe customer id should be numeric (1b8f824d5e)
- customer age shouldn't be an INT, because we don't get half ages
- is_chain should be, maybe, bool ???? (did not understood this variable)
- first order and last order should have the same type maybe
-customer age + first order + HR_0 should be integers, not floats
  


### Unique Values

In [8]:
#ABCDE.nunique()
#I put like a comment just because it was big 

In [9]:
ABCDE['last_promo'].unique()
# '-' not problematic, means there was no promo used

array(['DELIVERY', 'DISCOUNT', '-', 'FREEBIE'], dtype=object)

In [10]:
ABCDE['payment_method'].unique()
#Everyone payed?

array(['DIGI', 'CASH', 'CARD'], dtype=object)

In [11]:
ABCDE['customer_region'].unique()

array(['2360', '8670', '4660', '2440', '-', '4140', '2490', '8370',
       '8550'], dtype=object)

## 1.3 Basic Statistical Descriptions of Data

#### Initial Statistical overall analysis - Profile Report

In [20]:
from pandas_profiling import ProfileReport
pip instal 

Profile = ProfileReport(
     ABCDE,
     title='ABCDEeats Insigths',
     correlations={
         "pearson": {"calculate": True},
         "spearman": {"calculate": False},
         "kendall": {"calculate": False},
         "phi_k": {"calculate": False},
         "cramers": {"calculate": False},
     },
 

)

AttributeError: module 'pkgutil' has no attribute 'ImpImporter'

In [13]:
#Stats only for numeric types rounded with 2 decimal places
ABCDE.describe().round(2).T

count   mean    std   min   25%   50%    75%  \
customer_age              31161.0  27.51   7.16  15.0  23.0  26.0  31.00   
vendor_count              31888.0   3.10   2.77   0.0   1.0   2.0   4.00   
product_count             31888.0   5.67   6.96   0.0   2.0   3.0   7.00   
is_chain                  31888.0   2.82   3.98   0.0   1.0   2.0   3.00   
first_order               31782.0  28.48  24.11   0.0   7.0  22.0  45.00   
last_order                31888.0  63.68  23.23   0.0  49.0  70.0  83.00   
CUI_American              31888.0   4.88  11.65   0.0   0.0   0.0   5.66   
CUI_Asian                 31888.0   9.96  23.56   0.0   0.0   0.0  11.83   
CUI_Beverages             31888.0   2.30   8.48   0.0   0.0   0.0   0.00   
CUI_Cafe                  31888.0   0.80   6.43   0.0   0.0   0.0   0.00   
CUI_Chicken Dishes        31888.0   0.77   3.66   0.0   0.0   0.0   0.00   
CUI_Chinese               31888.0   1.43   8.19   0.0   0.0   0.0   0.00   
CUI_Desserts              31888.0   0.88   5.26   0.0   0.0   0.0   0.00   
CUI_Healthy               31888.0   0.95   5.83   0.0   0.0   0.0   0.00   
CUI_Indian                31888.0   1.63   7.44   0.0   0.0   0.0   0.00   
CUI_Italian               31888.0   3.23  11.25   0.0   0.0   0.0   0.00   
CUI_Japanese              31888.0   3.00  10.18   0.0   0.0   0.0   0.00   
CUI_Noodle Dishes         31888.0   0.71   4.54   0.0   0.0   0.0   0.00   
CUI_OTHER                 31888.0   3.00   9.77   0.0   0.0   0.0   0.00   
CUI_Street Food / Snacks  31888.0   3.91  15.55   0.0   0.0   0.0   0.00   
CUI_Thai                  31888.0   0.84   4.43   0.0   0.0   0.0   0.00   
DOW_0                     31888.0   0.56   1.01   0.0   0.0   0.0   1.00   
DOW_1                     31888.0   0.57   1.04   0.0   0.0   0.0   1.00   
DOW_2                     31888.0   0.59   1.05   0.0   0.0   0.0   1.00   
DOW_3                     31888.0   0.62   1.07   0.0   0.0   0.0   1.00   
DOW_4                     31888.0   0.68   1.09   0.0   0.0   0.0   1.00   
DOW_5                     31888.0   0.65   1.07   0.0   0.0   0.0   1.00   
DOW_6                     31888.0   0.70   1.17   0.0   0.0   0.0   1.00   
HR_0                      30723.0   0.00   0.00   0.0   0.0   0.0   0.00   
HR_1                      31888.0   0.05   0.32   0.0   0.0   0.0   0.00   
HR_2                      31888.0   0.06   0.35   0.0   0.0   0.0   0.00   
HR_3                      31888.0   0.12   0.50   0.0   0.0   0.0   0.00   
HR_4                      31888.0   0.10   0.44   0.0   0.0   0.0   0.00   
HR_5                      31888.0   0.08   0.36   0.0   0.0   0.0   0.00   
HR_6                      31888.0   0.07   0.33   0.0   0.0   0.0   0.00   
HR_7                      31888.0   0.08   0.38   0.0   0.0   0.0   0.00   
HR_8                      31888.0   0.13   0.64   0.0   0.0   0.0   0.00   
HR_9                      31888.0   0.23   0.72   0.0   0.0   0.0   0.00   
HR_10                     31888.0   0.33   0.89   0.0   0.0   0.0   0.00   
HR_11                     31888.0   0.38   0.96   0.0   0.0   0.0   0.00   
HR_12                     31888.0   0.31   0.84   0.0   0.0   0.0   0.00   
HR_13                     31888.0   0.24   0.64   0.0   0.0   0.0   0.00   
HR_14                     31888.0   0.22   0.60   0.0   0.0   0.0   0.00   
HR_15                     31888.0   0.28   0.74   0.0   0.0   0.0   0.00   
HR_16                     31888.0   0.36   0.87   0.0   0.0   0.0   0.00   
HR_17                     31888.0   0.39   0.94   0.0   0.0   0.0   0.00   
HR_18                     31888.0   0.34   0.89   0.0   0.0   0.0   0.00   
HR_19                     31888.0   0.25   0.80   0.0   0.0   0.0   0.00   
HR_20                     31888.0   0.14   0.59   0.0   0.0   0.0   0.00   
HR_21                     31888.0   0.07   0.35   0.0   0.0   0.0   0.00   
HR_22                     31888.0   0.05   0.30   0.0   0.0   0.0   0.00   
HR_23                     31888.0   0.05   0.28   0.0   0.0   0.0   0.00   


- Why first_order and customer_age is does not have a count of 31888? It is because of missing values? Yess it is
- Asian customers spend the most in average and they also have the highest maximum spend 


In [14]:
#Simple Stats including objets 
ABCDE.describe(include = "object")


customer_id customer_region last_promo payment_method
count        31888           31888      31888          31888
unique       31875               9          4              3
top     742ca068fc            8670          -           CARD
freq             2            9761      16748          20161

### Missing Values 

In [15]:
#check if there are missing values
ABCDE.isna().sum()

customer_id                    0
customer_region                0
customer_age                 727
vendor_count                   0
product_count                  0
is_chain                       0
first_order                  106
last_order                     0
last_promo                     0
payment_method                 0
CUI_American                   0
CUI_Asian                      0
CUI_Beverages                  0
CUI_Cafe                       0
CUI_Chicken Dishes             0
CUI_Chinese                    0
CUI_Desserts                   0
CUI_Healthy                    0
CUI_Indian                     0
CUI_Italian                    0
CUI_Japanese                   0
CUI_Noodle Dishes              0
CUI_OTHER                      0
CUI_Street Food / Snacks       0
CUI_Thai                       0
DOW_0                          0
DOW_1                          0
DOW_2                          0
DOW_3                          0
DOW_4                          0
DOW_5     

In [16]:
#Replacing the strings with NAN
ABCDE.replace("", np.nan, inplace=True)
#honestly didint see any changes , but I think that we don´t need to fix this now. Only if our grafs are weird.

In [17]:
#missing_values_per_column = ABCDE.isna().sum()


#f.plot_missing_values_bar(
    data=missing_values_per_column,
    xlabel='Columns',
    ylabel='Number of Missing Values',
    title='Missing Values per Column'
)


IndentationError: unexpected indent (1478143891.py, line 5)

#### Possible Solutions for mising Values 
- Customer_age -> fill with central tendency measure
- first_order -> replace with 0 if last_order=0
- HR_0 -> fill with sum(DOW_0:DOW_6) - sum(HR_1:HR_23)
- customer_region -> replace '-' with 9999

### Duplicates 

In [ ]:
ABCDE.duplicated().sum()
#we got 13 duplicates 

In [ ]:
ABCDE[ABCDE.duplicated()]

*Our Duplicates*
We got buyers from the same region, people with the same age. - Maybe this ones don't make sense to drop. dependent on our analysis. 
Question - customer_region don't appear in the sum of the duplicates but appears in this duplicated data frame

#### Metrics Division

#### Original ABCDE Corr

## Data Operations / Feature Enginering